cli/src/cmd/cast/logs.rs

```diff
@@ -0,0 +1,198 @@
+
+        let logs = cast.filter_logs(filter, json).await?;
+
+        println!("{}", logs);
+
+        Ok(())
+    }
+}
+
+fn build_filter(
+    from_block: Option<BlockNumber>,
+    to_block: Option<BlockNumber>,
+    address: Option<Address>,
+    sig_or_topic: Option<String>,
+    topics_or_args: Vec<String>,
+) -> Result<Filter, eyre::Error> {
+    let block_option = FilterBlockOption::Range { from_block, to_block };
+
+    let mut topics = match sig_or_topic {
+        Some(sig_or_topic) => match get_event(sig_or_topic.as_str()) {
+            Ok(event) => {
+                let args = topics_or_args.iter().map(|arg| arg.as_str()).collect::<Vec<_>>();
+
+                let indexed_inputs = event
+                    .inputs
+                    .iter()
+                    .zip(args)
+                    .filter(|(input, _)| input.indexed)
+                    .map(|(input, arg)| (&input.kind, arg))
+                    .collect::<Vec<_>>();
+
+                let indexed_tokens = parse_tokens(indexed_inputs, true)?;
+
+                let token0 = indexed_tokens.get(0);
+                let token1 = indexed_tokens.get(1);
+                let token2 = indexed_tokens.get(2);
+
+                let raw = match (token0, token1, token2) {
+                    (Some(token0), Some(token1), Some(token2)) => ethers::abi::RawTopicFilter {
+                        topic0: Topic::This(token0.clone()),
+                        topic1: Topic::This(token1.clone()),
+                        topic2: Topic::This(token2.clone()),
+                    },
+                    (Some(token0), Some(token1), None) => ethers::abi::RawTopicFilter {
+                        topic0: Topic::This(token0.clone()),
+                        topic1: Topic::This(token1.clone()),
+                        topic2: Topic::Any,
+                    },
+                    (Some(token0), None, None) => ethers::abi::RawTopicFilter {
+                        topic0: Topic::This(token0.clone()),
+                        topic1: Topic::Any,
+                        topic2: Topic::Any,
+                    },
+                    (None, None, None) => ethers::abi::RawTopicFilter {
+                        topic0: Topic::Any,
+                        topic1: Topic::Any,
+                        topic2: Topic::Any,
+                    },
+                    _ => panic!("Invalid number of indexed arguments"),
+                };
+
+                let filter = event.filter(raw).unwrap();
+                [filter.topic0, filter.topic1, filter.topic2, filter.topic3]
+                    .iter()
+                    .map(|topic| match topic {
+                        Topic::This(topic) => Some(ValueOrArray::Value(Some(*topic))),
+                        _ => None,
+                    })
+                    .collect::<Vec<_>>()
+            }
+            Err(_) => {
+                let mut topics = topics_or_args;
+                topics.reverse();
```

Comment:
Any suggestions on a more eloquent way to prepend an element to a potentially empty vector?